proposed/2022/dotnet-nuget-why-proposal.md

```diff
@@ -0,0 +1,131 @@
+
+![](../../meta/resources/TransitiveDependencies/DotNetCLI.png)
+
+## Motivation
+
+A developer should be able to understand where every package in a solution/project originated from. This can help them understand complex dependency chains they may have within their projects. Having this understanding can help a developer promote a transitive dependency quickly & easily to a top-level dependency in case there is a security concern or conflict. The `dotnet nuget why` command aims to help developers understand package dependency chains.
+
+## Explanation
+
+### Functional Explanation
+
+The NuGet restore operation generates a `project.assets.json` file under the `obj` folder for `PackageReference` style project. This file maintains a project's dependency graph. Therefore, this file can be used in order to produce a dependency graph when the `dotnet nuget why` command is called. 
+
+### Technical Explanation
+
+The `dotnet nuget why` command will print out the dependency graph of a given package only if it is part of the final graph.
+
+```
+dotnet nuget why [<PROJECT>|<SOLUTION>] <PACKAGE_NAME>
+    [-f|--framework <FRAMEWORK>]
+
+dotnet nuget why -h|--help
+```
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution is found, an error is thrown. If more than one project is found, the dependency graph for each project is printed. 
+
+- PACKAGE_NAME
+
+The package for whom the dependency graph has to be identified. Package name wildcards are not supported.
+
+#### Options
+
+- -f|--framework <FRAMEWORK>
+
+Search package dependency graph for a specific target framework.
+
+- -?|-h|--help
+
+Prints out a description of how to use the command.
+
+#### Examples
+
+- List dependency graph of a package given `package id`.
+
+```
+dotnet nuget why packageA
+
+Project 'projectNameA' has the following dependency graph for 'packageA'
+   [net6.0]: Microsoft.ML (1.0.0) -> Microsoft.ML.Util (1.0.0) -> packageA (1.0.0)
+   [net472] Microsoft.ML (1.0.0) -> Microsoft.ML.Util (1.0.0) -> packageA (1.0.0)
+
+Project 'projectNameB' has the following dependency graph for 'packageA'
+   [net6.0]: Microsoft.ML (1.1.0) -> Microsoft.ML.Util (1.1.0) -> packageA (1.1.0)
+   [net472] Microsoft.ML (1.1.0) -> Microsoft.ML.Util (1.1.0) -> packageA (1.1.0)
+```
+
+- List dependency graph of a package given `package id` when there is a diamond dependency (a package is brought in by more than one path).
+
+```
+dotnet nuget why packageA
+
+Project 'projectNameA' has the following dependency graph for 'packageA'
+   [net6.0]: Microsoft.ML (1.0.0) -> Microsoft.ML.Util (1.0.0) -> packageA (1.0.0)
+   [net6.0] Microsoft.ML (1.0.0) -> Microsoft.ML.SampleUtils (1.0.0) -> packageA (1.0.0)
```

Comment:
```suggestion
   [net6.0] Microsoft.ML (1.0.0) -> Microsoft.ML.SampleUtils (1.0.0) -> packageA (1.0.0)
```

Should net6.0 be a more of a top level thing and then we nest?

I don't have too strong of an opinion here. 

Ie. should we have:

```
   net6.0:
     Microsoft.ML (1.0.0) -> Microsoft.ML.Util (1.0.0) -> packageA (1.0.0)
     Microsoft.ML (1.0.0) -> Microsoft.ML.SampleUtils (1.0.0) -> packageA (1.0.0)
```

Reply:
Sure, this way it's not as redundant as it is right now.